File: 1481-least-number-of-unique-integers-after-k-removals/1481-least-number-of-unique-integers-after-k-removals.cpp
class Solution {
public:
    
    int findLeastNumOfUniqueInts(vector<int>& arr, int k) {
        unordered_map<int, int> mp;
        
        for(int i: arr){
            mp[i]++;
        }
        
        vector<int> vec;
        
        for(auto p: mp)
            vec.push_back(p.second);
        
        sort(begin(vec), end(vec));
        
        int ans = 0;
        int i=0;
        
        for(; i<vec.size(); i++){
            
            if(k>=vec[i]){
                k -= vec[i];
                vec[i]= 0;
            } else{
                vec[i] -= k;
                k=0;
                break;
            }
        }
        
        for(int j = i; j<vec.size(); j++){
            if(vec[i]) ans++;
        }
            
        
        return ans;
    }
};

File: 1481-least-number-of-unique-integers-after-k-removals/1481-least-number-of-unique-integers-after-k-removals.java
class Solution {
    public int findLeastNumOfUniqueInts(int[] arr, int k) {
        
        Map<Integer, Integer> hm = new HashMap<>();
        
        for(int i=0; i<arr.length; i++){
            hm.put(arr[i], hm.getOrDefault(arr[i], 0) + 1);
        }
        
        List<Integer> l = new ArrayList<>();
        
        for(Map.Entry<Integer, Integer> e: hm.entrySet()){
            l.add(e.getValue());
        }
        
        Collections.sort(l);
        
        for(int i=0; i<l.size(); i++){
            int val = l.get(i);
            if(k>=val){
                k -= val;
                l.set(i, 0);
            } else{
                l.set(i, val);
                k=0;
                break;
            }
        }
    
        int ans = 0;
        
         for(int i=0; i<l.size(); i++){
             if(l.get(i)>0) ans++;
         }
        
        return ans;
    }
}